File: src/main/java/sage/web/context/CommonModelAttributesHandler.java
package sage.web.context;

import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ModelAttribute;
import sage.domain.service.TagService;
import sage.domain.service.UserService;
import sage.transfer.UserSelf;
import sage.web.auth.Auth;

@ControllerAdvice("sage.web.page")
public class CommonModelAttributesHandler {
  @Autowired
  private UserService userService;
  @Autowired
  private TagService tagService;

  @ModelAttribute("userSelf")
  public UserSelf userSelf() {
    return Auth.cuidOpt().map(userService::getSelf).orElse(null);
  }

  @ModelAttribute("userSelfJson")
  public String userSelfJson() {
    return Auth.cuidOpt().map(cuid -> Json.json(userService.getSelf(cuid))).orElse(null);
  }

  @ModelAttribute("tagTreeJson")
  public String tagTreeJson() {
    return tagService.getTagTreeJson();
  }
}

File: src/main/kotlin/sage/web/context/CommonModelAttributesHandler.kt
package sage.web.context

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ModelAttribute
import sage.service.TagService
import sage.service.UserService
import sage.transfer.UserSelf
import sage.web.auth.Auth

@ControllerAdvice("sage.web.page")
open class CommonModelAttributesHandler @Autowired constructor(
    private val userService: UserService,
    private val tagService: TagService
) {
  private val log = LoggerFactory.getLogger(javaClass)

  @ModelAttribute("userSelf")
  open fun userSelf(): UserSelf? = Auth.uid()?.run { silent { userService.getSelf(this) } }

  @ModelAttribute("userSelfJson")
  open fun userSelfJson(): String? = Auth.uid()?.run { silent { Json.json(userService.getSelf(this)) } }

  @ModelAttribute("tagTreeJson")
  open fun tagTreeJson(): String? = silent { tagService.getTagTreeJson() }

  private inline fun <T> silent(f: () -> T) =
      try {
        f.invoke()
      } catch(e: Exception) {
        log.error("", e)
        null
      }
}
